model.py

```diff
@@ -280,7 +289,217 @@ def _restrict_to_stages(cls, qu, allowed_stages=None):
+        # by the subquery, in km. If a service area is "everywhere", the distance
+        # is 0.
+        def min_distance(subquery):
+            return func.min(
+                case(
+                    [(subquery.c.type==Place.EVERYWHERE, literal_column(str(0)))],
+                    else_=func.ST_Distance_Sphere(target, subquery.c.geometry)
+                )
+            ) / 1000
+
+        # Minimum distance to any eligibility area.
+        eligibility_min_distance = min_distance(eligibility_areas_subquery)
+
+        # Minimum distance to any focus area.
+        focus_min_distance = min_distance(focus_areas_subquery)
+
+        # Decrease the score based on how far away the library's eligibility area is.
+        score = score * exponential_decrease(1.0 * eligibility_area_distance_factor * eligibility_min_distance)
+ 
+        # Decrease the score based on how far away the library's focus area is.
+        score = score * exponential_decrease(1.0 * focus_area_distance_factor * focus_min_distance)
+
+        # Decrease the score based on the sum of the sizes of the library's focus areas, in km^2.
+        # This currently  assumes that the library's focus areas don't overlap, which may not be true.
+        # If a focus area is "everywhere", the size is the area of Earth (510 million km^2).
+        focus_area_size = func.sum(
+            case(
+                [(focus_areas_subquery.c.type==Place.EVERYWHERE, literal_column(str(510000000000000)))],
+                else_=func.ST_Area(focus_areas_subquery.c.geometry)
+            )
+        ) / 1000000
+        score = score * exponential_decrease(1.0 * focus_area_size_factor * focus_area_size)
+
+        # Rank the libraries by score, and remove any libraries
+        # that are below the score threshold.
+        library_id_and_score = select(
+            [libraries_collections.c.libraries_id,
+             score.label("score"),
+            ]
+        ).having(
+            score > literal_column(str(score_threshold))
+        ).where(
+            # Limit to the collection summaries for the user's language. If a library
+            # has no collection for the language, it's still included.
+            or_(
+                libraries_collections.c.collectionsummaries_language==language_code,
+                libraries_collections.c.collectionsummaries_language==None
+            )
+        ).select_from(
+            libraries_collections
+        ).group_by(
+            libraries_collections.c.libraries_id,
+            libraries_collections.c.collectionsummaries_id,
+            libraries_collections.c.collectionsummaries_size,
+        ).order_by(
+            score.desc()
+        )
+
+        result = _db.execute(library_id_and_score)
+        library_ids_and_scores = [r for r in result]
+        # Look up the Library objects and return them with the scores.
+        return [(get_one(_db, Library, id=lid), score) for lid, score in library_ids_and_scores]
```

Comment:
It would be better if the original query returned the Library objects, but I couldn't figure out a way to do that, since I had to use the SQL expression API instead of ORM.